src/main/scala/org/constellation/primitives/TrieBasedTipService.scala

```diff
@@ -0,0 +1,143 @@
+                          numFacilitatorPeers: Int,
+                          minPeerTimeAddedSeconds: Int)
+    extends ConcurrentTipService {
+
+  private val tips: TrieMap[String, TipData] = TrieMap.empty
+
+  override def set(newTips: Map[String, TipData]): Unit = {
+    tips ++= newTips
+  }
+
+  override def toMap: Map[String, TipData] = {
+    tips.toMap
+  }
+
+  def size: Int = {
+    tips.size
+  }
+
+  def get(key: String): Option[TipData] = {
+    tips.get(key)
+  }
+
+  def remove(key: String)(implicit metrics: Metrics): Unit = {
+    tips -= key
+    metrics.incrementMetric("checkpointTipsRemoved")
+    metrics.incrementMetric("checkpointTipsRemoved")
+  }
+
+  def update(checkpointBlock: CheckpointBlock)(implicit dao: DAO): Unit = {
+    // TODO: should size of the map be calculated each time or just once?
+    // previously it was static due to all method were sync and map updates
+    // were performed on the end
+//    def reuseTips: Boolean = tips.size < maxWidth
+    val reuseTips: Boolean = tips.size < maxWidth
+
+    checkpointBlock.parentSOEBaseHashes.distinct.foreach { h =>
+      tips.get(h).foreach {
+        case TipData(block, numUses) if !reuseTips || numUses >= 2 =>
+          remove(block.baseHash)(dao.metrics)
+        case TipData(block, numUses) if reuseTips && numUses <= 2 =>
+          dao.metrics.incrementMetric("checkpointTipsIncremented")
+          put(block.baseHash, TipData(block, numUses + 1))(dao.metrics)
+      }
+    }
+
+    put(checkpointBlock.baseHash, TipData(checkpointBlock, 0))(dao.metrics)
+  }
+
+  def put(k: String, v: TipData)(implicit metrics: Metrics): Option[TipData] = {
+    if (tips.size < sizeLimit) {
+      tips.put(k, v)
+    } else {
+      // TODO: should newest override oldest cache-like? previously was just discarded
+//      thresholdMetCheckpoints = thresholdMetCheckpoints.slice(0, 100)
+      metrics.incrementMetric("memoryExceeded_thresholdMetCheckpoints")
+      metrics.updateMetric("activeTips", tips.size.toString)
+      None
+    }
+  }
+
+  def getMinTipHeight()(implicit dao: DAO): Long =
+    tips.keys
+      .map {
+        dao.checkpointService.get
+      }
+      .flatMap {
+        _.flatMap {
+          _.height.map {
+            _.min
+          }
+        }
+      }
+      .min
+
+  override def pull(
+    readyFacilitators: Map[Id, PeerData]
+  )(implicit metrics: Metrics): Option[(Seq[SignedObservationEdge], Map[Id, PeerData])] = {
+
+    metrics.updateMetric("activeTips", tips.size.toString)
+
+    (tips.size, readyFacilitators) match {
+      case (x, facilitators) if x >= 2 && facilitators.nonEmpty =>
+        val tipSOE = calculateTipsSOE()
+        Some(tipSOE -> calculateFinalFacilitators(facilitators, tipSOE.foldLeft("")(_ + _.hash)))
+      case (x, _) if x >= 2 =>
+        Some(calculateTipsSOE() -> Map.empty[Id, PeerData])
+      case (_, _) => None
+    }
+  }
+
+  private def calculateTipsSOE(): Seq[SignedObservationEdge] = {
+    Random
```

Comment:
This method could potentially be a real performance killer. It's called every time we pull and involves turning the entire map into a sequence, shuffling the entire collection, only to take 2 elements.

I'm fine with it for now, but if it ever becomes a problem (esp. if we decide to let tips grow fairly large), we may want to rethink his somehow.

Reply:
Ok, it was copy paste from previous one. I fixed it by checking if it's less than magic number 50 and slicing it otherwise, what do you think ?